Fix table regressions.

jasmussen opened the pull request:
This PR fixes table block regressions introduced in https://github.com/WordPress/gutenberg/pull/7899#issuecomment-411732824

The block property for the table makes sure that it's mobile responsive. The "display table" is necessary for fixed layout to work.

- `display: table;` on `tbody` is necessary for the `has-fixed-layout` option to work. Please see https://github.com/WordPress/gutenberg/pull/6314 for a GIF
- `display: block` on `.wp-block-table` is necessary for the table itself to be responsive. Please see https://github.com/WordPress/gutenberg/pull/2152.

Note that the responsiveness for the table is in the `theme.scss` file, which means themes have to opt into these "opinionated styles".

Please test this PR with and without the fixed table cells option enabled, frontend and backend. And please verify that this PR does not regress what #7899 was meant to fix.

Commit:
Fix table regressions.

This PR fixes table block regressions introduced in https://github.com/WordPress/gutenberg/pull/7899#issuecomment-411732824

The block property for the table makes sure that it's mobile responsive. The "display table" is necessary for fixed layout to work.

iseulde commented on packages/block-library/src/table/theme.scss:
```diff
@@ -1,5 +1,6 @@
 .wp-block-table {
 	overflow-x: auto;
+	display: block; // Necessary in order for this table to be mobile responsive.
```
This seems odd to me, changing `display` of a table element. I had some problem with this rule in a  PR, so I took it out. What Exactly does mobile responsive mean here? Could you elaborate?

jasmussen replied:
To answer your question, if we had the whole thing wrapped in a div, we could do this instead:

https://codepen.io/joen/pen/RYojVZ?editors=1100

But as the markup is today, we need to leverage the `table` as the full-width block to hold the horizontal scrollbar, and then make a child element of that (in this case, tbody/thead/tfoot) be the full-wide table that has a min-width.

iseulde commented on packages/block-library/src/table/style.scss:
```diff
@@ -1,6 +1,7 @@
 .wp-block-table {
 	// Fixed layout toggle
 	&.has-fixed-layout tbody {
+		display: table;
```
Should this be added as well for `thead` and `tfoot`?

iseulde replied:
I think this should definitely be resolved.

iseulde commented on packages/block-library/src/table/style.scss:
```diff
@@ -1,6 +1,7 @@
 	// Fixed layout toggle
 	&.has-fixed-layout tbody {
+		display: table;
 		table-layout: fixed;
```
Why is this not used on the `table` element? This seems a bit odd.

jasmussen replied:
Same reason as noted in https://github.com/WordPress/gutenberg/pull/9306#discussion_r213971788 essentially.

iseulde commented:
Wondering if there could be any alternative fix for unresponsiveness. I'm guessing tables are by nature unresponsive though... If you have a few columns, you'll want to scroll them, not have them all squashed together. I wouldn't mess with the normal CSS rules of a table to make it more responsive. Could it help to wrap the table in a (scrollable) `figure` tag or something?

earnjam commented:
I think I had found a fix to this issue, but was holding off because I wasn't sure if everything was going to be redone as part of #6923. I'll see if I can find it on a local branch and compare to what @iseulde merged in @8767.

jasmussen commented:
All good feedback, thank you.

As a preface, there are two features of the table block being discussed. One is the fixed layout, which is off by default and toggled on in the sidebar.

The other is the responsiveness, which, note, is stored in the `theme.scss` file and therefore something a theme has to opt into, to receive. 

But the idea here is that every block should come with a minimum level of mobile responsiveness. And a "true" responsive table isn't really a table anymore, it's a rather complex collection of divs put together to look like a table. This is completely fair to explores separately, but as the vanilla offering we have to consider the basics which are accessible.

For that reason, the most minimal responsiveness I could find, was to _have the table scroll horizontally, if less than 360px wide_. You can see that here: https://codepen.io/joen/pen/zJqNgj?editors=1100

Due to the nature of how the `display: table;` property works, though, this involves some trickery to make the full width behave more like a div than a table, because the table will always try to contain its contents.

Here's a GIF:

![responsive table](https://user-images.githubusercontent.com/1204802/44646572-c43c0580-a9db-11e8-8fec-0a5fd19e4ee7.gif)

If we can accomplish a minimally responsive table in a different way, I'd love suggestions.

iseulde commented:
I think the thead and tfoot should also be added. I feel a bit uncomfortable with changing the display of a table from the default, but if this is valid and the way to do it, ok. Is there any way tests could be added for this (unsure here)? This seems like something with so many different configurations that we could easily break in the future.

jasmussen commented:
I could use some help getting this PR in ship shape if you have bandwidth. I feel strongly that all blocks we ship should have some measure of responsiveness, remember it's opt-in for themes. 

Right now both the responsiveness and the fixed-cell-width are broken in master, so it would be nice to find some solution soon, even if we need to refine it later.

iseulde commented:
@jasmussen Cool, I'm fine with just restoring the display rule. Could we just add `thead` and `tfoot` with the `tbody` rule to display as a table as well? That makes sense right? Then it seems good to go.

jasmussen commented:
Will do, feel free to separately track any concerns you have and we'll look at it. 👍 👍

Commit:
Address feedback: include thead, tfoot

jasmussen commented:
Addressed the feedback. Thanks for the reviews.

If we need to refactor the table display values to not re-do the tbody/tfoot/thead elements, we need a wrapping div, see https://codepen.io/joen/pen/RYojVZ?editors=1100

gziolo commented:
![screen shot 2018-08-30 at 14 34 01](https://user-images.githubusercontent.com/699132/44851804-c57e6400-ac61-11e8-9df8-f957351c86c5.png)

this is what I see when I create a new table

gziolo commented:
It seems like styles on the frontend are flipped:

![screen shot 2018-08-30 at 14 38 11](https://user-images.githubusercontent.com/699132/44852006-5f461100-ac62-11e8-9c63-65cdf3f19f6b.png)

![screen shot 2018-08-30 at 14 38 46](https://user-images.githubusercontent.com/699132/44852023-6ff68700-ac62-11e8-99eb-31d44276f0c1.png)

jasmussen committed:
Fix collapsed on init table.

jasmussen commented:
Hmm, you're right. And this suggests @iseulde was right to be nervous as well, because what causes the table to collapse like that is the changing to `block` display. 

I pushed a fix in https://github.com/WordPress/gutenberg/pull/9306/commits/b7d41bf344c7e5f5a84c3855a26484fbca43ce86, and I do think we should merge this because it fixes a regression — it is frustrating that the fixed width switch is broken in master.

But if we discover that changing the display properties of the table, thead, tbody, and tfoot, we should look at an alternative fix, which I think would include a wrapping div, see: https://codepen.io/joen/pen/RYojVZ. 

What do you think? Should we get this merged so we can fix the regression?

jasmussen commented:
Oh, taking a look at the inverted styles in the Twenty theme.

jasmussen committed:
Fix fixed-width cells.

jasmussen commented:
Fix the fixed-width option in Twenty. Good catch.

![screen shot 2018-08-30 at 15 37 40](https://user-images.githubusercontent.com/1204802/44855239-edbe9080-ac6a-11e8-9ea3-8e23299233dd.png)

![screen shot 2018-08-30 at 15 37 44](https://user-images.githubusercontent.com/1204802/44855243-eeefbd80-ac6a-11e8-86b5-9fb931b2578e.png)

![screen shot 2018-08-30 at 15 37 51](https://user-images.githubusercontent.com/1204802/44855246-f0b98100-ac6a-11e8-857d-e50ababc96ed.png)

![screen shot 2018-08-30 at 15 39 04](https://user-images.githubusercontent.com/1204802/44855250-f1eaae00-ac6a-11e8-9d52-8794a39f1b3b.png)

jasmussen commented:
Thanks, let's get this in and look at any other enhancements separately.

talldan commented:
> But if we discover that changing the display properties of the table, thead, tbody, and tfoot, we should look at an alternative fix, which I think would include a wrapping div

It looks like this is the cause of https://github.com/WordPress/gutenberg/issues/9779.

thead, tbody and tfoot need to be set to their default style properties to work. We then also have the secondary issue that the table needs to be `display: table` for `table-layout: fixed;` to work.

Unfortunately it looks like we need another approach. Probably another `div` as you say @jasmussen.

Seeing as we're discussing it on #9801, I think I'll introduce a fix there. That PR also handily introduces a block deprecation, so probably makes sense to handle it there.
